Add COPY command

swamp0407 opened the pull request:
I'm not used to using git and I'm not familiar with how to do PR, so I'm sorry if I made a mistake.

Syntax:

`COPY <key> <new-key> [DB <dest-db>] [REPLACE]`

see #6599 

Co-authored-by: tmgauss

I'm not very good at English, so there may be something wrong with the grammar of the comments in the code, etc. I'm sorry, but please check.

madolson commented on src/server.c:
```diff
@@ -622,6 +622,10 @@ struct redisCommand redisCommandTable[] = {
      0,NULL,1,1,1,0,0,0},
 
+    {"copy",copyCommand,-3,
+     "write fast @keyspace",
```
+use-memory

swamp0407 replied:
thanks @madolson ! I've fixed the flag in the command. I also fixed the copy command a bit.

But the code doesn't pass the test.
What do I need to do to make sure my code passes the test?

Also, should I combine my commits into one?

oranagra replied:
i guess `fast` should be removed? considering some types may require O(N) loops?

also, not sure about `@keyspace`
the only hint in the docs is:
```
 * 7) The non obvious "keyspace" category includes the commands
 *    that interact with keys without having anything to do with
 *    specific data structures, such as: DEL, RENAME, MOVE, SELECT,
 *    TYPE, EXPIRE*, PEXPIRE*, TTL, PTTL, ...
```
so maybe i'd consider COPY similar to BRPOPLPUSH?
i guess not (we should keep `@keyspace`), but wanted to raise this concern to see if anyone has any feedback.

madolson replied:
I think we should keep @keyspace

madolson commented on src/db.c:
```diff
@@ -1070,6 +1070,102 @@ void moveCommand(client *c) {
+    long long dbid, expire;
+    int j, replace = 0;
+    if (server.cluster_enabled) {
+        addReplyError(c,"COPY is not allowed in cluster mode");
```
I don't think we got strong conensus on this, but I think you should be able to copy in cluster mode, you just can't copy between DBs.

madolson reviewed (commented):
I'll take a deeper look through all the copying tomorrow but I added some high level comments. 

I think I fixed the test with this PR: https://github.com/redis/redis/pull/7966. Redis ACLs are being simplified in weird ways.

madolson commented on src/db.c:
```diff
@@ -1070,6 +1070,104 @@ void moveCommand(client *c) {
+    for (j = 3; j < c->argc; j++) {
+        if (!strcasecmp(c->argv[j]->ptr,"replace")) {
+            replace = 1;
+        } else if (getLongLongFromObject(c->argv[j],&dbid) == C_ERR ||
```
This isn't the same as the syntax as well. You should also be looking for the keyword DB, so "COPY key key2 DB 5 or something. This allows us to be more future proof in adding more arguments here.

madolson commented on tests/unit/keyspace.tcl:
```diff
@@ -169,6 +169,72 @@ start_server {tags {"keyspace"}} {
+        assert {[r get mynewkey] eq "foobar"}
+        r flushdb
+    }
+
```
It would be nice to add a test for all the different types. You can use debug digest to validate that the key/values are the same.

madolson replied:
Looks like the stream digest doesn't include consumer group information, probably worth expanding on that.

oranagra replied:
i'm not certain we want digest to include it.
there are commands that modify these which consider them metadata.
IIRC they do perform `dirty++` and are propagated to replicas and RDB/AOF, and only avoid doing signalModifyKey (invalidating WATCH), so maybe that means that they should indeed be in the digest.
@guybe7 anything to add?

guybe7 replied:
since CGs go in the AOF and RDB i think they should be included in the digest

oranagra commented:
just mentioning that i didn't go over the code yet (just took a quick look at the comments @madolson gave). 

p.s. maybe in a followup PR, you can add module API callback so that modules can provide cloning functionality. but let's put that aside for now.

madolson commented on src/db.c:
```diff
@@ -1070,6 +1070,104 @@ void moveCommand(client *c) {
+    }
+
+    /* Check if the element exists and get a reference */
+    o = lookupKeyWrite(c->db,key);
```
lookupKeyRead

oranagra replied:
actually, for some unclear reason, redis uses lookupKeyWrite when reading keys on behalf of a write operation. looking at sunionstore, this intent is clear.
see https://github.com/redis/redis/issues/7475

madolson replied:
Ok, point taken, it's weird behavior but good to know.

madolson commented on src/object.c:
```diff
@@ -211,6 +211,361 @@ robj *dupStringObject(const robj *o) {
+    quicklistSetOptions(lobj->ptr, server.list_max_ziplist_size, server.list_compress_depth);
+    llen = listTypeLength(o);
+    listTypeIterator *iter = listTypeInitIterator(o, start, LIST_TAIL);
+    while (llen--) {
```
A quicklist is a linked list of ziplists, which are binary blobs. You should be able to copy each blob and string them together in a quicklist, it'll probably be much faster than this.

madolson commented on src/object.c:
```diff
@@ -211,6 +211,361 @@ robj *dupStringObject(const robj *o) {
+            break;
+    }
+    llen = zsetLength(o);
+    if (zobj->encoding == OBJ_ENCODING_ZIPLIST) {
```
Same comment as before, ziplists are binary blocks in an allocation, you can just copy the whole thing.

madolson commented on src/db.c:
```diff
@@ -1070,6 +1070,104 @@ void moveCommand(client *c) {
+        {
+            addReply(c,shared.outofrangeerr);
+            return;
+        }
```
Would prefer to error instead of silently ignoring bad arguments.

madolson commented on src/object.c:
```diff
@@ -211,6 +211,361 @@ robj *dupStringObject(const robj *o) {
+            streamCG *new_cg = streamCreateCG(new_s, (char *)ri_cgroups.key,
+                                              ri_cgroups.key_len, &cg->last_id);
+            /* If already exists */
+            if (new_cg == NULL) {
```
This shouldn't be possible? Would prefer to assert here.

madolson commented on src/object.c:
```diff
@@ -211,6 +211,361 @@ robj *dupStringObject(const robj *o) {
+            argv[j * 2] = createObject(OBJ_STRING, sdsnewlen((const char *)key, key_len));
+            argv[j * 2 + 1] = createObject(OBJ_STRING, sdsnewlen((const char *)value, value_len));
+        }
+        streamAppendItem(new_s, argv, numfields, &id, &id);
```
I was going to suggest building a helper method for this that just adds an sds array directly without converting them to robjs. That would be nice, but streamAppendItem() is a sprawling function. You could also see if you can do the same optimization and just copy the list packs.

oranagra replied:
i agree, let's avoid going over the individual elements and just run over the rax and copy the listpacks.

madolson commented on src/stream.h:
```diff
@@ -113,10 +113,12 @@ streamCG *streamLookupCG(stream *s, sds groupname);
 void streamPropagateConsumerCreation(client *c, robj *key, robj *groupname, sds consumername);
-
+int streamAppendItem(stream *s, robj **argv, int64_t numfields, streamID *added_id, streamID *use_id);
+robj *createObjectFromStreamID(streamID *id);
```
Still needed?

madolson commented on src/db.c:
```diff
@@ -1070,6 +1070,104 @@ void moveCommand(client *c) {
+    notifyKeyspaceEvent(NOTIFY_GENERIC,"copy_to",c->argv[2],dst->id);
+
+    server.dirty++;
+    addReply(c,shared.cone);
```
Not sure how I feel about this specifically, but we maybe should retain the LRU information. We could also document the fact we just don't do that.

oranagra replied:
we recently made an effort to replace many of the dbDelete+dbAdd pairs with setKey (see SUNIONSTORE and alike).
The difference is that `setKey` retains the LRU of the destination key.

However, my above note is actually unrelated since we're talking of the LRU of the source key.
from what i can tell, both MOVE and RENAME, retain the LRU of the source key, so i'm guessing COPY should do the same?

Or if we consider that a new object (unlike MOVE and RENAME), we should behave similarly to RESTORE REPLACE (create a fresh new key)?
or maybe this is in some way similar to SUNIONSTORE being called with one input (clones a set and keeps the destination LRU)?

@redis/core-team @guybe7 please share your thoughts. (i'm leaning towards the RESTORE REPLACE behavior)

guybe7 replied:
i vote for:
src's LRU should remain unchanged (like MOVE)
dst's LRU should be changed (like RESTORE REPLACE)

madolson commented on src/object.c:
```diff
@@ -211,6 +211,361 @@ robj *dupStringObject(const robj *o) {
+    }
+    streamIteratorStop(&si);
+
+    if (s->cgroups == NULL) {
```
```
if (s->cgroups == NUL) return sobj;
```

Keeps the indentation down.

oranagra replied:
good idea about indentation.
just a nit pick about the `if`, i prefer actions that change the execution flow (goto, break, return), to be on a line of their own (unless they come in an heavily repeated pattern)

madolson commented on src/object.c:
```diff
@@ -211,6 +211,361 @@ robj *dupStringObject(const robj *o) {
+        /*Nothing to do*/
+    } else {
+        /* Consumer Groups */
+        raxIterator ri_cgroups;
```
@guybe7 Do you agree that it makes sense to copy all of the consumer group information when we copy the stream. It seems weird to me, but it is metadata on the stream but also isn't directly data about the stream. Do you have any thoughts?

guybe7 replied:
@madolson yes, i think CGs and consumers should be copied as well... as i see it, COPY should provide a perfect copy of the key, i.e. executing the same command on both keys should have identical result

madolson reviewed (changes requested):
Most of it reviewed now! Generally looks good.

swamp0407 commented:
thanks @madolson @oranagra @guybe7 !
I fixed the code, but I wasn't sure how to fix things like removing the fast flag and keeping the LRU.
Please review.

oranagra commented:
@swamp0407 you did remove the `fast` flag (added `use-memory` instead).
and i think the conclusion from the discussion was that we want it to behave like RESTORE, so we we use dbDelete and dbAdd (don't keep the LRU).
both of these aspects are fine in the last version of the code in that respect (unless we hear other concerns / opinions that support changing that).
I see you changed to lookupKeyRead, which is wrong IMHO, let's wait for @madolson to respond in that discussion.

madolson commented:
The lookupKeyRead behavior is weird, but it looks like it's well established in other places, so I guess we should keep that consistent. Still a couple of other open comments. @swamp0407

oranagra commented on src/object.c:
```diff
@@ -211,6 +211,316 @@ robj *dupStringObject(const robj *o) {
+        zskiplistNode *ln;
+        sds ele;
+
+        /* Add a score-element pair to a new zset object which encoding is SKIPLIST. */
+        ln = zsl->header->level[0].forward;
+        while (llen--) {
+            ele = ln->ele;
+            zsetAdd(zobj, ln->score, ele, &retflags, NULL);
+            ln = ln->level[0].forward;
+        }
```
i recently remembered that zslInsert (used by zsetAdd) is more efficient when iterating on the list backwards: 
https://github.com/redis/redis/blob/f6546eff452e9868569610c797765db8f56763e3/src/rdb.c#L855-L861

it may be a topic for another PR to improved zslInsert, but maybe in the meanwhile we wanna run backwards here too (and put a comment saying why)?

swamp0407 commented:
thanks @oranagra @madolson 
In addition to the areas you suggested I should fix, I have also changed the code that copies the inset object.

Should I also do a fix to include the consumer group information in the digest of the stream key? I'm not sure I can implement it correctly.

oranagra commented on src/object.c:
```diff
@@ -211,6 +211,303 @@ robj *dupStringObject(const robj *o) {
+        newis->encoding = is->encoding;
+        newis->length = is->length;
+        zfree(set->ptr);
+        set->ptr = newis;
```
i don't like this intimate code that messes with the intset internals being in object.c
i.e. `intrev32ifbe(is->length)`

so i would have argued that this block should be promoted to some `intsetDup` function that will be implemented in intset.c and called here.
but in fact, there's simply a better, easier way to do do.

just call `intsetBlobLen` then malloc and memcpy the entire thing without bothering to understand what's inside.

oranagra commented on src/object.c:
```diff
@@ -211,6 +211,303 @@ robj *dupStringObject(const robj *o) {
+        unsigned char *zl = o->ptr;
+        size_t sz = ziplistBlobLen(zl);
+        unsigned char *new_zl = zmalloc(sz);
+        memcpy(new_zl, zl, sz);
```
i see you did this for ziplist, why didn't you do that for intset?

oranagra commented on src/object.c:
```diff
@@ -211,6 +211,303 @@ robj *dupStringObject(const robj *o) {
+            break;
+        case OBJ_ENCODING_HT:
+            hobj = createHashObject();
+            hashTypeConvert(hobj, OBJ_ENCODING_HT);
```
it's a good idea to call dictExpand here too, so that it'll be created with the right size and not go though rehashing.
same in the other paces that we create a dict which we know it's final size

oranagra replied:
use `dictSize(d)` instead of `d->ht[0].size`. it's both nicer not to access the internals of the dict, but also more "correct" since if the source dict is in the process of rehashing, it's records are split between ht[0] and ht[1].

oranagra commented on src/db.c:
```diff
@@ -1070,6 +1070,107 @@ void moveCommand(client *c) {
+                dbid < INT_MIN || dbid > INT_MAX ||
+                selectDb(c, dbid) == C_ERR)
+            {
+                addReplyErrorFormat(c, "-NODB No such number of db '%s' ",(char*)c->argv[j+1]->ptr);
```
i don't think we want to invent a new "NODB" error code.
probably best to reply with the same "invalid DB index" that's used in SELECT"

oranagra reviewed (commented):
haven't yet reviewed the streams code.

i think the fix for DEBUG DIGEST and steams can be a separate PR.
unless this one really needs it in order to validate correct result.
maybe you can use XINFO FULL instead for now.

swamp0407 commented:
Thank you for reviewing my code. @oranagra 

Please check it out as I have corrected it.
I have also fixed my mistake that I found in the dupStream function.

And, I don't need to use the stream consumer group's digest, so I won't touch it.

oranagra commented on src/object.c:
```diff
@@ -211,6 +211,304 @@ robj *dupStringObject(const robj *o) {
+            zobj = createZsetZiplistObject();
+            break;
+        case OBJ_ENCODING_SKIPLIST:
+            zobj = createZsetObject();
```
we should better use `dictExpand` here too (see that rdb.c does it also).

oranagra commented on src/object.c:
```diff
@@ -211,6 +211,304 @@ robj *dupStringObject(const robj *o) {
+        ln = zsl->tail;
+        while (llen--) {
+            ele = ln->ele;
+            zsetAdd(zobj, ln->score, ele, &retflags, NULL);
```
speaking of rdb.c, i see that rdb.c does this:
```
            znode = zslInsert(zs->zsl,score,sdsele);
            dictAdd(zs->dict,sdsele,&znode->score);
```
and you used `zsetAdd` which seems to be a top level function (used mainly for the command implementation: `zaddGenericCommand`)

i think it's better to use zslInsert and dictAdd directly in this case.

oranagra reviewed (commented):
few more comments..
btw, it may be a good idea to move all the 'dupXxxxObject` functions to their respective type c files eventually.
all this type specific logic sounds like it shouldn't be in object.c.
but let's do that after we're done with the other comments, for an easier review of everyone involved.

oranagra commented on src/object.c:
```diff
@@ -211,6 +211,304 @@ robj *dupStringObject(const robj *o) {
+                streamDecodeID(ri_cpel.key, &id);
+                streamEncodeID(buf, &id);
+                /* Insert NACK. */
+                streamNACK *new_nack = zmalloc(sizeof(*new_nack));
+                new_nack->delivery_time = nack->delivery_time;
+                new_nack->delivery_count = nack->delivery_count;
+                new_nack->consumer = new_consumer;
+                int group_inserted =
+                    raxTryInsert(new_cg->pel, buf, sizeof(buf), new_nack, NULL);
+                int consumer_inserted = raxTryInsert(
+                    new_consumer->pel, buf, sizeof(buf), new_nack, NULL);
+
+                if (group_inserted == 0) {
+                    streamFreeNACK(new_nack);
+                    new_nack = raxFind(new_cg->pel, buf, sizeof(buf));
+                    serverAssert(new_nack != raxNotFound);
+                    raxRemove(new_nack->consumer->pel, buf, sizeof(buf), NULL);
+                    /* Update the consumer and NACK metadata. */
+                    new_nack->delivery_time = nack->delivery_time;
+                    new_nack->delivery_count = nack->delivery_count;
+                    new_nack->consumer = new_consumer;
+                    /* Add the entry in the new consumer local PEL. */
+                    raxInsert(new_consumer->pel, buf, sizeof(buf), new_nack,
+                              NULL);
+                } else if (group_inserted == 1 && consumer_inserted == 0) {
+                    serverPanic("NACK half-created. Should not be possible.");
+                }
```
IIUC what this code does it:
1. create a NACK
2. insert it to the group and the consumer
3. if the group insertion failed (since the group already has one):
	1. delete the newly created nack
	2. locate the the group nack
	3. remove the newly added consumer nack
	4. add the group nack into the consumer.

to me this seems like a complicated and wrong approach, the more trivial one IMHO is:
1. search the group, if it already has one:
	1. add it to the consumer
2. else:
	1. create a new nack
	3. add to both the group and consumer

advantage:
- more logical approach IMHO (and probably shorter too)
- not allocating a nack just in order to free it soon.
- involves 3 rax searches max (1 find and 2 insertions), existing approach has 5 (2 insertions, find, removal and insertion)

oranagra replied:
looking at your recent change, i don't think that's good.
the PEL record in the consumer is expected to be the same pointer as the one in the consumer group PEL.
i see your recent code just calls `raxTryInsert` and `raxInsert`, so if the same record exists in two consumers, one will be shared with the group, and the other one will not (memory leak later).

however, i see that raxTryInsert and raxInsert can return the existing pointer if there's a clash.
so you can use that.
this will mean that you'll have just two searches (not 3 as in my previous proposal), but only other hand you'll have excessive calls to malloc / free, i suppose that's not so bad, and the code will be cleaner.

swamp0407 replied:
thanks @oranagra
I'm sorry but, I didn't understand exactly how to fix it, so I took a look at rdb.c to fix it.
Would this fix be a problem?

oranagra replied:
@swamp0407 new code looks good. (being the same approach that rdb loading takes, it can't be wrong)
also, it seems i got confused, one nack record can't exist in two consumers.

swamp0407 commented:
thanks @oranagra 
I have corrected my codes.

I wrote the code in reference to streamReplyWithRange in t_stream.c, but I didn't understand the streamReplyWithRange function correctly, so the part about adding nack to the group and consumer pel got complicated.

After reading the code again, I thought that the same key nack would never go into multiple people's pel.

I also fixed the part about creating a new consumer.
Let me know if there is  any problems with this fix.

oranagra commented:
@swamp0407 looks like i [responded](https://github.com/redis/redis/pull/7953#discussion_r519395397) to your last push before you submitted your last message.
AFAIK the entries in the PEL of the group and consumers point to the same memory (see rdb loading).

madolson commented on src/stream.h:
```diff
@@ -113,10 +113,12 @@ streamCG *streamLookupCG(stream *s, sds groupname);
 streamConsumer *streamLookupConsumer(streamCG *cg, sds name, int flags, int *created);
 streamCG *streamCreateCG(stream *s, char *name, size_t namelen, streamID *id);
 streamNACK *streamCreateNACK(streamConsumer *consumer);
+void streamEncodeID(void *buf, streamID *id);
```
Not used

swamp0407 replied:
done

madolson commented on src/stream.h:
```diff
@@ -113,10 +113,12 @@ streamCG *streamLookupCG(stream *s, sds groupname);
 void streamFreeNACK(streamNACK *na);
 void streamIncrID(streamID *id);
 void streamPropagateConsumerCreation(client *c, robj *key, robj *groupname, sds consumername);
+int streamAppendItem(stream *s, robj **argv, int64_t numfields, streamID *added_id, streamID *use_id);
```
Not used

swamp0407 replied:
done

madolson reviewed (commented):
Two minor things, but besides that LGTM.

itamarhaber commented:
Before we seal this one, I have a couple of notes:

* Should this be a multi-key command? I'm torn between liking the simplicity (e.g. `MOVE`), my general dislike of multi-key commands but acknowledging there may be a future (see `MIGRATE`).
* Should this support a `DELETE` (or similar) option to delete the source key? This would make this into an uber-command that could eventually replace/deprecate `MOVE` and `RENAME`
* The current implementation copies the source's TTL, if set. I agree with this default, but perhaps we'd like to add a `NOTTL` (or similar) option.

oranagra commented:
1. it must be a multi-key command, unless we only want to let it copy a key to another db (keeping it's name), which i think is a big miss
2. i would have considered it a good idea to add a DELETE argument if they where sharing code, but the fact is that if we do that, it'll just be a shortcut to the MOVE/RENAME implementation which is a `fast` command. so i'm not sure that's wise. but anyway, we can always add more arguments in the future.
3. i guess we can add `NOTTL`, `EX <ttl>`, and even `KEEPTTL`. not sure if that's needed, but either way we can add later.

itamarhaber commented:
WRT 1, I guess I meant variadic keys, something like `COPY SOURCE key [...] DEST key [...]`

oranagra commented:
ohh... 
i don't see the value of that (just sending it inside a MULTI will do the same thing, no disadvantages).
has anyone ever requested such a thing for MOVE or RENAME?

madolson commented:
I don't like the multi-key approach when there are flags involved, it really complicates the command. As oran said you can always multi-exec them. I also agree that we can add flags later as needed. 

From my understanding the follows up here are:
1. Reorganize the functions so that they are located in their respective types.
2. Extend the debug digest functionality so that it also includes consumer groups.

Since neither of these require a major consensus, I am OK with the current iteration. @redis/core-team conensus?

itamarhaber commented on src/db.c:
```diff
@@ -1070,6 +1070,107 @@ void moveCommand(client *c) {
+        };
+    }
+
+    if (delete) {
+        dbDelete(dst,newkey);
+    }
```
Ref: the `REPLACE` logic in `RESTORE`

```suggestion
    if (delete && dbDelete(dst,newkey)) {
        notifyKeyspaceEvent(NOTIFY_GENERIC,"del",c->argv[2],dst->id);
    }
```

oranagra replied:
@itamarhaber are you sure?
the "del" notification in restore is only in case the new key is expired, and restore deleted the old key.
it's something i recently copied from EXPIREAT (when ttl is in the past).

in the case restore succeeds it only does:
```
    notifyKeyspaceEvent(NOTIFY_GENERIC,"restore",key,c->db->id);
```

taking another example, SINTERSTORE it does:
```
   notifyKeyspaceEvent(NOTIFY_SET,"sinterstore", dstkey,c->db->id);
```
or (only in case a key was deleted, and the stored key is empty):
```
   notifyKeyspaceEvent(NOTIFY_GENERIC,"del", dstkey,c->db->id);
```

so i think in this case we just want the "copy_to" notification, not the "del" one

itamarhaber replied:
@oranagra You are of course correct - the `del` isn't needed here.

oranagra commented on src/db.c:
```diff
@@ -1070,6 +1070,107 @@ void moveCommand(client *c) {
+        } else if (!strcasecmp(c->argv[j]->ptr, "db") && additional >= 1) {
+            if (getLongLongFromObject(c->argv[j+1], &dbid) == C_ERR ||
+                dbid < INT_MIN || dbid > INT_MAX ||
+                selectDb(c, dbid) == C_ERR)
```
this trick of using `selectDb` to check if the dbid is in range is not very good.
if we exit with syntax error on the next argument, the selected db remains the dest db.
either re-select the original db right away, or just match the dbid range manually.

oranagra reviewed (commented):
i went over all the comments to check that we didn't forget anything, found another minor issue.

other than that:
1. revert the additional "del" keyspace notification
2. and move all the "dupXxxx" functions to their respective C files. try to find the correct place in each C file for this kind of API (not part of the commands section), and make sure to give them the right prefix. please do this refactory in a separate commit, so it'll be easy to review actual changes.

after that, i guess we're ready to merge this.

oranagra commented:
@swamp0407 thank you. merged.
